tools/gn

```diff
@@ -212,19 +238,39 @@ def to_gn_args(args):
       # The GN arg is not available in the windows toolchain.
       gn_args['enable_lto'] = enable_lto
 
-    if args.target_os:
+    # Set OS, CPU arch for host or target build.
+    if is_host_build(args):
+      gn_args['host_os'] = get_host_os()
+      gn_args['host_cpu'] = get_host_cpu(args)
+      gn_args['target_os'] = gn_args['host_os']
+      gn_args['target_cpu'] = get_target_cpu(args)
+      gn_args['dart_target_arch'] = gn_args['target_cpu']
+    else:
       gn_args['target_os'] = args.target_os
-    gn_args['target_cpu'] = get_target_cpu(args)
-    gn_args['dart_target_arch'] = gn_args['target_cpu']
+      gn_args['target_cpu'] = get_target_cpu(args)
+      gn_args['dart_target_arch'] = gn_args['target_cpu']
 
-    # No cross-compilation on Windows (for now).
-    # See: https://github.com/flutter/engine/pull/3883
+    # No cross-compilation on Windows (for now). Use host toolchain that
+    # matches the bit-width of the target architecture.
     if sys.platform.startswith(('cygwin', 'win')) and args.target_os != 'win':
+      gn_args['host_cpu'] = cpu_for_target_arch(gn_args['target_cpu'])
       gn_args['target_cpu'] = cpu_for_target_arch(gn_args['target_cpu'])
 
-    if args.target_os == 'ios':
+    # macOS host builds (whether x64 or arm64) must currently be built under
+    # Rosetta on Apple Silicon Macs.
+    # TODO(cbracken): https://github.com/flutter/flutter/issues/103386
+    if is_host_build(args) and gn_args['host_os'] == 'mac':
+      gn_args['host_cpu'] = 'x64'
+
+    # macOS target builds (whether x64 or arm64) must currently be built under
+    # Rosetta on Apple Silicon Macs.
+    # TODO(cbracken): https://github.com/flutter/flutter/issues/103386
+    if 'target_os' in gn_args and gn_args['target_os'] == 'mac':
+      gn_args['host_cpu'] = 'x64'
```

Comment:
These three blocks (Windows, macOS host builds, macOS target builds) are broken out separately since they're likely to have separate fixes.